notebook2/notebook/notebook.py

```diff
@@ -188,6 +198,131 @@ def external_url_for(path):
     return url + path
 
 
+# Kube has max 63 character limit
+def user_id_transform(user_id): return hashlib.sha224(user_id.encode('utf-8')).hexdigest()
+
+
+def read_svc_status(svc_name: str):
+    try:
+        # TODO: inspect exception for non-404
+        _ = k8s.read_namespaced_service(svc_name, 'default')
+        return 'Running'
+    except Exception:
+        return 'Deleted'
+
+
+def read_containers_status(container_statuses):
+    """
+        Summarize the container status based on its most recent state
+
+        Parameters
+        ----------
+        container_statuses : list[V1ContainerStatus]
+            https://github.com/kubernetes-client/python/blob/master/kubernetes/docs/V1ContainerStatus.md
+    """
+    if container_statuses is None:
+        return None
+
+    state = container_statuses[0].state
+    rn = None
+    wt = None
+    tm = None
+    if state.running:
+        rn = {"started_at": state.running.started_at}
+
+    if state.waiting:
+        wt = {"reason": state.waiting.reason}
+
+    if state.terminated:
+        tm = {"exit_code": state.terminated.exit_code, "finished_at": state.terminated.finished_at,
+              "started_at": state.terminated.started_at, "reason": state.terminated.reason}
+
+    if rn is None and wt is None and tm is None:
+        return None
+
+    return {"running": rn, "terminated": tm, "waiting": wt}
+
+
+def read_conditions(conds):
+    """
+        Return the most recent status=="True" V1PodCondition or None
+
+        Parameters
+        ----------
+        conds : list[V1PodCondition]
+            https://github.com/kubernetes-client/python/blob/master/kubernetes/docs/V1PodCondition.md
+
+    """
+    if conds is None:
+        return None
+
+    maxDate = None
+    maxCond = None
+    for condition in conds:
+        if maxDate is None:
+            maxCond = condition
+            maxDate = condition.last_transition_time
+            continue
+
+        if condition.last_transition_time > maxDate and condition.status == "True":
+            maxCond = condition
+            maxDate = condition.last_transition_time
+
+    return {
+        "message": maxCond.message,
+        "reason": maxCond.reason,
+        "status": maxCond.status,
+        "type": maxCond.type}
```

Comment:
This method can be:
```
conds = [c.to_dict for c in conds if c.status == "True"]
if not conds:
    return None
latest = max(conds, key=lambda x: x.last_transition_time)
return latest.to_dict()
```

Reply:
So this function is slightly different, in two ways than the one your wrote. First, in the case that status is "False", and is the only status, I return it. This allows us to state we're "Waiting for: " + condition in the ui. Second, there are more values returned by V1PodCondition than I return here. And, a minor note, in my version, we iterate only 1 list, and create 0 lists. 

Given this, do you still prefer yours?

Reply:
Ah. I think that logic is more clearly described by:
```
if conds is None:
    return conds
return max(conds, key=lambda x: (x.last_transition_time, x.status)).to_dict()
```